core/trino-main/src/main/java/io/trino/operator/exchange/LocalExchange.java

```diff
@@ -306,11 +306,11 @@ private static PartitionFunction createPartitionFunction(
                 bucketToPartition);
     }
 
-    public static int getBucketCount(Session session, NodePartitioningManager nodePartitioningManager, PartitioningHandle partitioning)
+    public static int getBucketCount(Session session, NodePartitioningManager nodePartitioningManager, PartitioningHandle partitioning, int partitionCount)
     {
         if (partitioning.getConnectorHandle() instanceof MergePartitioningHandle) {
             // TODO: can we always use this code path?
-            return nodePartitioningManager.getNodePartitioningMap(session, partitioning).getBucketToPartition().length;
+            return nodePartitioningManager.getNodePartitioningMap(session, partitioning, partitionCount).getBucketToPartition().length;
```

Comment:
nit: consider adding comment why it's ok to pass arbitrary `partitionCount` here